### PR TITLE
New: Refactor Extra Files for Movies

### PR DESCRIPTION
**Pull request by Qstick**
#### Database Migration
YES

#### Description
- Adds Support for Import of Existing Metadata/Subtitle/Other Extra Files
- Adds Support for Metadata File and Image Generation (poster, banner, fanart)
- Move Metadata on movie folder changes
- Add Option to XBMC to write nfo to <filename>.nfo or movie.nfo
- Add Ability to find current images with either format <filename>-poster.ext or poster.ext 
- Look for cdart, logo, landscape, banner, fanart, poster as existing options.
- Refactor DeleteBadMediaCovers Housekeeper to work with movies.
- Trigger MovieRenamedEvent on rename so that extras and notifications are triggered.

#### Todos
- [x] Tests
- [x] API Endpoint for Extra Files
- [x] UI Showing Extra Files

#### Issues Fixed or Closed by this PR
#121 
#167 

Should also close #2262 and close #1104

### CODE REVIEW

**Comment by Qstick** [comment]
@galli-leo make sure I didn't miss anything or break anything, Its late here :)

**Comment by Qstick** [comment]
> Did you try out what happens if you have auto rename folders & dynamic paths?

From what I can tell its working as expected. I tested a few scenarios:
- Static Path with movie file and no metadata. Metadata is created as expected after a movie refresh depending on which options are selected
- Change movie file naming and trigger a rename, metadata that is based on the moviefile name (ex. \<movieFile>-poster.jpg) gets renamed properly when the movie file is renamed and the DB gets updated with the new naming.
- Static Path with movie file and existing metadata/subtitles. Metadata is recognized by Radarr and scanned into DB. If the extra file matches one of the consumers formats it is added to the MetadataFiles table, if extra file matches the subtitle extensions it is added to the SubtitleFiles table, and if metadata doesn't match either its considered an "other" extra file and is added to the ExtraFiles table.
- After both of the above scenarios I changed the folder path naming format and edited that movie to be dynamic instead of static. When a organize or refresh is triggered the move and all metadata gets moved correctly.

>I assume that's why you removed the code where Sonarr compares old & new path for extra file?
>Do you think that we should pull that code out from that function?

Which particular bit are you referring to here? 

Sonarr had 5 metadata types (Series Metadata, Series Images, Season Images, Episode Metadata, Episode Images). I took the assumption that with Movies we would only have two (Movie Images, Movie Metadata). That's where a lot of the code purge came into play. 

The one thing that I did remove that we may want to look for another solution to was building out metadata when the folder is created, but no file exists. I took the approach that all movie metadata was file related given nfo files or image files are generally named based on the moviefile name (ex. \<movieFile>.mkv, \<movieFile>.nfo, \<movieFile>-poster.jpg). That being said I know that Kodi and others also support movie.nfo for metadata and poster.jpg, banner.jpg, etc... for images. Currently those are recognized as valid metadata files during import, but it's not generating metadata until the movie file is present.  I didn't really want to have 4 metadata types (MovieMetadata, MovieFile Metadata, MovieImages, MovieFileImages) because they are essentially the same. Thoughts on this?

>Also do you think it would be difficult to add another table to the files tab with all extra files?

Good Idea I'll work on this.

File: src/NzbDrone.Core/Extras/Metadata/Consumers/MediaBrowser/MediaBrowserMetadata.cs [hunk 1]
```diff
@@ -25,76 +25,63 @@ public class MediaBrowserMetadata : MetadataBase<MediaBrowserMetadataSettings>
...
-                tvShow.Add(new XElement("Network", series.Network));
-                tvShow.Add(new XElement("Airs_Time", series.AirTime));
+                movieElement.Add(new XElement("id", movie.ImdbId));
+                movieElement.Add(new XElement("Status", movie.Status));
 
-                if (series.FirstAired.HasValue)
-                {
-                    tvShow.Add(new XElement("FirstAired", series.FirstAired.Value.ToString("yyyy-MM-dd")));
-                }
+                movieElement.Add(new XElement("ContentRating", movie.Certification));
```
Comment (galli-leo) [inline src/NzbDrone.Core/Extras/Metadata/Consumers/MediaBrowser/MediaBrowserMetadata.cs]: Not sure, but I don't think we use the Certification field (i.e. we don't save it).

File: src/NzbDrone.Core/Extras/Metadata/Consumers/MediaBrowser/MediaBrowserMetadata.cs [hunk 2]
```diff
@@ -25,76 +25,63 @@ public class MediaBrowserMetadata : MetadataBase<MediaBrowserMetadataSettings>
...
-                tvShow.Add(new XElement("ProductionYear", series.Year));
-                tvShow.Add(new XElement("RunningTime", series.Runtime));
-                tvShow.Add(new XElement("IMDB", series.ImdbId));
-                tvShow.Add(new XElement("TVRageId", series.TvRageId));
-                tvShow.Add(new XElement("Genres", series.Genres.Select(genre => new XElement("Genre", genre))));
-
-                var persons   = new XElement("Persons");
-
-                foreach (var person in series.Actors)
+                foreach (var person in movie.Actors)
```
Comment (galli-leo) [inline src/NzbDrone.Core/Extras/Metadata/Consumers/MediaBrowser/MediaBrowserMetadata.cs]: Same for actors, I think we don't store them.

File: src/NzbDrone.Core/Extras/Others/OtherExtraService.cs [hunk 3]
```diff
@@ -15,71 +15,41 @@ namespace NzbDrone.Core.Extras.Others
...
-                            _diskProvider.MoveFile(existingFileName, newFileName);
-                            extraFile.RelativePath = series.Path.GetRelativePath(newFileName);
-                            movedFiles.Add(extraFile);
-                        }
-                        catch (Exception ex)
-                        {
-                            _logger.Warn(ex, "Unable to move extra file: {0}", existingFileName);
-                        }
-                    }
+                    movedFiles.AddIfNotNull(MoveFile(movie, movieFile, extraFile));
```
Comment (galli-leo) [inline src/NzbDrone.Core/Extras/Others/OtherExtraService.cs]: Sorry should have been more specific with the renaming code. I meant this bit here. Sonarr seems to rename and move the metadata file.

**Comment by galli-leo** [comment]
> Which particular bit are you referring to here?
See comment above.

> The one thing that I did remove that we may want to look for another solution to was building out metadata when the folder is created, but no file exists. I took the approach that all movie metadata was file related given nfo files or image files are generally named based on the moviefile name (ex. <movieFile>.mkv, <movieFile>.nfo, <movieFile>-poster.jpg). That being said I know that Kodi and others also support movie.nfo for metadata and poster.jpg, banner.jpg, etc... for images. Currently those are recognized as valid metadata files during import, but it's not generating metadata until the movie file is present. I didn't really want to have 4 metadata types (MovieMetadata, MovieFile Metadata, MovieImages, MovieFileImages) because they are essentially the same. Thoughts on this?
Sounds good, I don't think anybody needs metadata before a file is present, since you won't use it in kodi anyways.

File: src/NzbDrone.Core/Extras/Others/OtherExtraService.cs (hunk 3 above)
Comment (Qstick) [reply src/NzbDrone.Core/Extras/Others/OtherExtraService.cs]: This bit of code was just not eliminated, but moved up to the base class `ExtraFileManager` to eliminate duplication same as upstream.

**Comment by Qstick** [comment]
@galli-leo API and UI added and comments addressed. Let me know if you see anything else.

**Comment by galli-leo** [comment]
Thanks a ton for being patient with my UI wishes, lol

**Comment by Buster14** [comment]
Thanks for this guys! just one small request, would it be possible to somehow set the poster name into just poster.jpg instead of moviename.poster.jpg? I have an apps that is set to read poster.jpg.. also i thought by default kodi reads poster.jpg?

**Comment by Qstick** [comment]
@Buster14 I can add a option for this similar to nfo option. Kodi reads both, but having the movie title in the name allows for all files in single movie folder (should radarr ever support that), so I felt that was the more preferable default.

**Comment by Buster14** [comment]
Strange my kodi doesn't detect it, i'm using older Kodi version cause newer version breaks my custom skin... Could you possibly add movie.jpg to the next version? I think it's better to make radarr more customizable for users.. thanks before

**Comment by Buster14** [comment]
Just noticed now It's using poster.jpg instead of moviename-poster, thanks alot! radarr rocks

**Comment by Buster14** [comment]
Hi guys, i have another question, my library is sorted based on western and asian. 
And this requires "production_countries" from tmdb api, cause unfortunately "language" properties wont help cause most of them are english.

I will then use ps script to auto map the movie from a temp folder. 

So i was wondering if its possible to include this as well in movie.nfo?

**Pull request merged**